lightwood/mixers/helpers/plinear.py

```diff
@@ -0,0 +1,171 @@
+
+# Import PyTorch
+import torch #
+import torch.nn as nn # import modules
+
+from torch.nn.parameter import Parameter # import Parameter to create custom activations with learnable parameters
+import torch.nn.functional as F # import torch functions
+from torch.nn import init
+import math
+import random
+import numpy as np
+
+
+
+class PLinear(nn.Module):
+    '''
+    Implementation of probabilistic weights via Linear function
+    '''
+    def __init__(self, in_features, out_features, bias=True):
+        """
+
+        :param in_features:  as name suggests
+        :param out_features: this essentially the number of neurons
+        :param bias: if you want a specific bias
+        """
+
+        super(PLinear ,self).__init__()
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+        # these are the matrices that we will optimize for
+        self.sigma = Parameter(torch.Tensor(out_features, in_features))
+        self.mean = Parameter(torch.Tensor(out_features, in_features))
+
+        # there can be various ways to sample, given various distributions, we will stick with discrete normal as it is way faster
+        self.w_sampler = self.w_discrete_normal
+
+        if bias:
+            self.bias = Parameter(torch.Tensor(out_features))
+        else:
+            self.register_parameter('bias', None)
+
+        self.reset_parameters()
+
+        #make sure that we tell the graph that these two need to be optimized
+        self.sigma.requiresGrad = True # set requiresGrad to true!
+        self.mean.requiresGrad = True  # set requiresGrad to true!
+
+
+    def reset_parameters(self):
+        """
+        This sets the initial values for the distribution parameters, mean, sigma
+
+        """
+        if self.w_sampler == self.w_normal:
+            init.kaiming_uniform_(self.mean, a=math.sqrt(5)) # initial means (just as in original linear)
+            init.uniform_(self.sigma, a=0.01, b=0.1) # initial sigmas from 0.5-2
+        elif self.w_sampler == self.w_discrete_normal:
+            init.kaiming_uniform_(self.mean, a=math.sqrt(5))  # initial means (just as in original linear)
+            init.uniform_(self.sigma, a=0.05, b=0.2)
+
+        if self.bias is not None:
+            fan_in, _ = init._calculate_fan_in_and_fan_out(self.mean)
+            bound = 1 / math.sqrt(fan_in)
+            init.uniform_(self.bias, -bound, bound)
+
+    def w_discrete_normal(self):
+        """
+        Sample a w matrix based on a discrete normal distribution
+        :return: w
+        """
+
+        bucket_seed = random.uniform(0,1) # gett a number that can help determines which bucket of the distribution can numbers fall in
```

Comment:
The code from here to line 81 could be replaced with:

`sigma_multiplier = np.random.choice([1,2,3],p=[0.33,0.47,0.2])`

and it would probably make it easier to understand and shorter.